### ISSUE TITLE
config: add a flag to skip blocksync if num(peers)=0 and num(vals) > 0

### ISSUE THREAD

**Issue by czarcas7ic** [cometbft/cometbft#1679]
## Bug Report

Many times, we create state exported testnets based on mainnet state. Usually, these networks don't start unless fast sync is set to true, at which point they start. Before knowing this hack, we used "kicker nodes", or nodes we simply peered to the validator to get the network to start. When moving over to comet BFT, I found that the fast sync hack no longer works, and "kicker nodes" are once again required. Removing the following lines prevents the need for a kicker node or the fast sync hack https://github.com/cometbft/cometbft/blob/6302491554b2fa895091aeb99c448d7fc4ca0781/internal/blocksync/pool.go#L172-L176

I am not sure if just removing those lines is the best way forward, I am just pointing out that there is once again large overhead in creating these state exported testnets, as multiple nodes are once again required. Please let me know if any further information is needed.

**Comment by melekes** [comment]
hey @czarcas7ic 👋 thanks for reporting this.

> these networks don't start unless fast sync is set to true, at which point they start

There's no option to set fast sync to true. Did you mean [state syncs](https://docs.cometbft.com/v0.38/core/state-sync)? Does that mean they were state syncing to the latest state?

> I found that the fast sync hack no longer works, 

Why? Are there any log messages?

> Removing the following lines prevents the need for a kicker node or the fast sync hack

So the node neither [state syncs](https://docs.cometbft.com/v0.38/core/state-sync) nor [block syncs](https://docs.cometbft.com/v0.38/core/block-sync), correct?

**Comment by melekes** [comment]
It looks like you're using CometBFT `v0.34`. I'd recommend upgrading to the latest version (`v0.38`) as soon as possible.

> When moving over to comet BFT, I found that the fast sync hack no longer works, and "kicker nodes" are once again required.

What's the target CometBFT version here? v0.34, v0.37, v0.38? Please note that in the last two releases, `fast_sync` config option was renamed to `blocksync`.

```
[blocksync]

# Block Sync version to use:
#
# In v0.37, v1 and v2 of the block sync protocols were deprecated.
# Please use v0 instead.
#
#   1) "v0" - the default block sync implementation
version = "v0"
```

**Comment by melekes** [comment]
Closing due to no response. @czarcas7ic please comment here if you still need any help. thanks 🙏

**Comment by czarcas7ic** [comment]
Many other chains on v0.34 will suffer from this issue, and the change I recommended is state compatible. 

The node has zero peers because it is the only node on the testnet

**Comment by czarcas7ic** [comment]
For now we are likely going to just run on a cometbft fork to get past this issue, but I imagine the issue is going to still be there when we upgrade and will have to fork again.

**Comment by jmalicevic** [comment]
Hi, as @melekes  said, in newer versions fast_sync (which is renamed to blocksync) is not optional anymore and therefore there is no config. But regardless, I am trying to understand your scenario, you have one node with a full blockstore up to height H. You want to immediately start consensus from height H and that did not work without blocksync enabled? There is a [check](https://github.com/cometbft/cometbft/blob/5aa9fa1a71d0f8f636a646980205141edb2a99b4/node/node.go#L812) in Comet atm verifying that the [node in question is not the only validator](https://github.com/cometbft/cometbft/blob/5aa9fa1a71d0f8f636a646980205141edb2a99b4/node/node.go#L363). If it is, then blocksync is skipped. I imagine this check returns false for you because it is an exported state where you indeed had many validators. If the state and block stores have all the data, I am wondering what parameters is blocksync setting for you (when this check is removed), that the consensus replay is not able to set without blocksync.

**Comment by czarcas7ic** [comment]
Can this issue be re-opened as well? Osmosis has lost days of work due to this overhead, and Stride can also attest to this being problematic.

**Comment by sampocs** [comment]
+1 to keeping this issue opened

The state exported testnets are one of the most important parts of our (Stride's) upgrade testing flow and this would make the process much less painful!

**Comment by melekes** [comment]
@czarcas7ic got it, thanks 🙏

**Comment by andynog** [comment]
I believe this is what [ADR-104 Out of Band state sync](https://github.com/cometbft/cometbft/blob/main/docs/references/architecture/adr-104-out-of-band-state-sync.md) does.

But there is no implementation and I believe got de-prioritized. Some [context](https://github.com/cometbft/cometbft/issues/27)

**Comment by czarcas7ic** [comment]
Hey, this is still causing a lot of pain for us in our testing. Would someone be able to explain how they want this issue addressed so I can make a PR addressing it?

**Comment by p0mvn** [comment]
+1 on this. The node just does not start with:
```
block_sync = true
```

Toggling to false helps. We are on v0.37. If this setting is removed in the future but enabled by default with no other changes, we will not be able to productively operate testnets.

**Comment by czarcas7ic** [comment]
I think I found a way around this for future testnets. Right now, cometbft checks `isOnlyValidatorUs`. The issue is, if we do a state exported testnet and leave the validators there in the store, despite us being the only ACTIVE validator, this check returns as false because the validators are still present. If there is a way for this check to instead determine if we are the only actively signing validator that would be ideal, but for now if I delete every validator from the store I can get the network to start without the `block_sync` work around.

**Comment by melekes** [comment]
@czarcas7ic @p0mvn hey all! can you check if https://github.com/cometbft/cometbft/pull/2512 fixes your problem? Thanks 🙏

**Comment by czarcas7ic** [comment]
Hey @melekes, we were able to get around the problem by ensuring we delete all references to all other validators but our own. I reviewed the logic though and this seems sound for chains that have yet to move over to the new in-place-testnet and still rely on state exports!

**Comment by melekes** [comment]
Closing in favor of https://github.com/cometbft/cometbft/issues/2610
